feat(medusa): Extend Module Resolution configuration

pKorsholm opened the pull request:
**What**
update module definitions in `ConfigModule` to support the following module configuration methods:
- none: defaults are loaded
- boolean `inventoryService: true|false`, if true the defaults are loaded, false throws if it's a required module, if the module is not required it's not loaded 
- string: `inventoryService: "..."`, treats the string as a path to the overriding module, (we dont handle the case where a string is given but the module is not overridable, we just load the default in that case)
- `ConfigurableModuleDeclaration = { resolve?: string, options?: Record } `, like plugins, options can be used to pass configs to the main service of the module, if not defined no options are passed. Resolve is like string, if defined it's used to look for a custom module, otherwise the default is loaded.

Testing: 
- Added unit tests for: 
  - `loaders/modules.ts`
  - `loaders/module-definitions/index.ts`

Fixes CORE-932

srindom commented on packages/medusa/src/loaders/module.ts:
```diff
@@ -20,22 +20,36 @@ export default async ({
     try {
+      if (!resolution.resolutionPath) {
+        container.register({
+          [resolution.definition.registrationName]: asFunction(
+            () => false
+          ).singleton(),
```
nit:
```suggestion
          [resolution.definition.registrationName]: asValue(false).singleton(),
```

srindom commented on packages/medusa/src/types/global.ts:
```diff
@@ -71,7 +77,7 @@ export type ConfigModule = {
   }
   featureFlags: Record<string, boolean | string>
-  modules?: Record<string, string>
+  modules?: Record<string, boolean | string | ConfigurableModuleDeclaration>
```
nit: we will never allow `true` so we could change to:
```suggestion
  modules?: Record<string, false | string | ConfigurableModuleDeclaration>
```

srindom commented on packages/medusa/src/loaders/module-definitions/index.ts:
```diff
@@ -10,15 +10,36 @@ export default ({ modules }: ConfigModule) => {
+      }
+      if (!mod) {
+        moduleResolutions[definition.key] = {
+          resolutionPath: undefined,
```
nit: could delete

olivermrbl commented on packages/medusa/src/loaders/index.ts:
```diff
@@ -132,9 +132,15 @@ export default async ({
   const servicesActivity = Logger.activity(`Initializing services${EOL}`)
   track("SERVICES_INIT_STARTED")
-  servicesLoader({ container, configModule, isTest })
+  await servicesLoader({ container, configModule, isTest })
```
todo: Should not be awaited

srindom commented on packages/medusa/src/loaders/module.ts:
```diff
@@ -20,22 +20,36 @@ export default async ({
+        continue
+      }
+
       const loadedModule = await import(resolution.resolutionPath!)
```
thought: can this fail? e.g. if the user has not specified anything but hasn't installed the `defaultPackage` in this scenario import would not be able to get the package. It would be helpful to show an error message with:

```
We could not resolve [resolutionPath] for module [moduleKey]. You may need to run yarn add [resolutionPath].
```

olivermrbl commented on packages/medusa/src/loaders/module-definitions/index.ts:
```diff
@@ -10,15 +10,36 @@ export default ({ modules }: ConfigModule) => {
   for (const definition of MODULE_DEFINITIONS) {
     let resolutionPath = definition.defaultPackage
 
+    const mod = projectModules[definition.key]
```
todo: Change `mod` to `module`

srindom replied:
think module is a reserved keyword

olivermrbl replied:
Ah yeah, might be. Would do `projectModule` then :)

pKorsholm replied:
not reserved, but I have changed it to `moduleConfiguration` anyways to avoid confusion (even vscode changed the color of the variable-name thinking it was a keyword).

srindom reviewed (approved):
LGTM w. comments 💪

olivermrbl commented on packages/medusa/src/loaders/module-definitions/definitions.ts:
```diff
@@ -1,5 +1,22 @@
 
-export const MODULE_DEFINITIONS: ModuleDefinition[] = []
+export const MODULE_DEFINITIONS: ModuleDefinition[] = [
+  // {
+  //   key: "stockLocationService",
+  //   registrationName: "stockLocationService",
+  //   defaultPackage: "@medusajs/stock-locations",
+  //   label: "StockLocationService",
+  //   isRequired: false,
+  //   canOverride: true,
+  // },
+  // {
+  //   key: "inventoryService",
+  //   registrationName: "inventoryService",
+  //   defaultPackage: "@medusajs/inventory",
+  //   label: "InventoryService",
+  //   isRequired: false,
+  //   canOverride: true,
+  // },
```
todo: rm

olivermrbl reviewed (commented):
LGTM, just a couple of suggestions :)

olivermrbl commented on packages/medusa/src/loaders/module.ts:
```diff
@@ -46,15 +60,22 @@ export default async ({
     } catch (err) {
       if (resolution.definition.isRequired) {
         throw new Error(
-          `Could not resolve required module: ${resolution.definition.label}`
+          `Could not resolve required module: ${resolution.definition.label}.`
         )
```
thought: Instead of throwing a generic error, we could throw the error from the module loader? I've added that in the event bus module PR, so would love to hear what you think. Would make it easier to debug :)

olivermrbl replied:
So do:
```ts
   if (resolution.definition.isRequired) {
        logger.warn(`Could not resolve required module: ${resolution.definition.label}`)
        throw err // throw error from module loader on required modules
      }
```

pKorsholm commented:
@srindom @olivermrbl
I have updated the module loader and refactored the individual registration into a method that returns `void | error`. If we return an error a warning is raised in the console and if the module is required the error is thrown.

It was the most flexible solution I could come up with on the spot without throwing if statements everywhere to find out if we are going to throw or just log 😅

srindom reviewed (approved):
LGTM! Pending @olivermrbl
